### ISSUE TITLE
no-invalid-this false positive for method declaration syntax

### ISSUE THREAD

**Issue by amacleay** [palantir/tslint#4033]
### Bug Report

- __TSLint version__: 5.10.0
- __TypeScript version__: 2.9.2
- __Running TSLint via__: (pick one) CLI 

#### TypeScript code being linted

```ts
function bindProperty(this: { someProperty: number }, fn: Function) {
  return fn.bind(this);
}

const anotherObj: Record<string, Function> = {
  bindProperty: function(this: { someProperty: number }, fn: Function) {
    return fn.bind(this);
  },
};

const someObj: Record<string, Function> = {
  bindProperty(this: { someProperty: number }, fn: Function) {
    return fn.bind(this);
  },
};
```

with `tslint.json` configuration:

```json
{
    "defaultSeverity": "error",
    "jsRules": {},
    "rules": {
      "no-invalid-this": [true]
    },
    "rulesDirectory": []
}
```

#### Actual behavior
Even though `someObj` and `anotherObj` produce identical javascript output because one is just a shorthand for the other, the method declaration style produces a tslint error
```$ tslint --project . --format stylish
/tmp/tmp.DcFq9uCAcp/index.ts:13:20
ERROR: 13:20  no-invalid-this  the "this" keyword is disallowed outside of a class body

```

#### Expected behavior
All of these examples should pass lint

#### Full repro example

`for x in $( git ls-files | grep -v yarn ); do echo;echo "# $x"; cat $x; done`

```

# .gitignore
*.js
node_modules

# index.ts
function bindProperty(this: { someProperty: number }, fn: Function) {
  return fn.bind(this);
}

const anotherObj: Record<string, Function> = {
  bindProperty: function(this: { someProperty: number }, fn: Function) {
    return fn.bind(this);
  },
};

const someObj: Record<string, Function> = {
  bindProperty(this: { someProperty: number }, fn: Function) {
    return fn.bind(this);
  },
};

# package.json
{
  "name": "tmp.dcfq9ucacp",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "UNLICENSED",
  "devDependencies": {
    "tslint": "^5.10.0",
    "typescript": "^2.9.2"
  }
}

# tsconfig.json
{
  "compilerOptions": {
    "target": "ES5",
    "strict": true,
    "strictNullChecks": true
  }
}

# tslint.json
{
    "defaultSeverity": "error",
    "jsRules": {},
    "rules": {
      "no-invalid-this": [true]
    },
    "rulesDirectory": []
}
```

**Comment by NHuebner1983** [comment]
Not sure if this is helpful, but it's giving a false positive for lambda functions:

```    
public updateGeneralInformation = async (userInfo: IUpdateGeneralInformation): Promise<AxiosResponse> =>
        this.asyncPatch('/general', userInfo);
```

This method is inside of a class extending a class that carries the public `asyncPatch` method.

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
